# Copyright (c) 2020  Carnegie Mellon University
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

lifecycle_manager_local_navigation:
  ros__parameters:
    use_sim_time: True
    bond_timeout: 30.0
    bond_connect_timeout: 30.0
    bond_heartbeat_period: 1.0
    get_state_timeout: 30.0
    change_state_timeout: 30.0

bt_navigator:
  ros__parameters:
    bond_timeout: 30.0
    bond_connect_timeout: 30.0
    bond_heartbeat_period: 1.0
    use_sim_time: True
    global_frame: local/odom
    robot_base_frame: local/base_footprint
    odom_topic: /local/odom
    enable_groot_monitoring: False
    plugin_lib_names:
    - nav2_compute_path_to_pose_action_bt_node
    - nav2_compute_path_through_poses_action_bt_node
    - nav2_follow_path_action_bt_node
    - nav2_back_up_action_bt_node
    - nav2_spin_action_bt_node
    - nav2_wait_action_bt_node
    - nav2_clear_costmap_service_bt_node
    - nav2_is_stuck_condition_bt_node
    - nav2_goal_reached_condition_bt_node
    - nav2_goal_updated_condition_bt_node
    - nav2_initial_pose_received_condition_bt_node
    - nav2_reinitialize_global_localization_service_bt_node
    - nav2_rate_controller_bt_node
    - nav2_distance_controller_bt_node
    - nav2_speed_controller_bt_node
    - nav2_truncate_path_action_bt_node
    - nav2_goal_updater_node_bt_node
    - nav2_recovery_node_bt_node
    - nav2_pipeline_sequence_bt_node
    - nav2_round_robin_node_bt_node
    - nav2_transform_available_condition_bt_node
    - nav2_time_expired_condition_bt_node
    - nav2_distance_traveled_condition_bt_node
    - nav2_single_trigger_bt_node
    - nav2_is_battery_low_condition_bt_node
    - nav2_navigate_through_poses_action_bt_node
    - nav2_navigate_to_pose_action_bt_node
    - nav2_remove_passed_goals_action_bt_node
    - nav2_planner_selector_bt_node
    - nav2_controller_selector_bt_node
    - nav2_goal_checker_selector_bt_node
    - cabot_people_exist_bt_node
    - cabot_someone_on_path_bt_node
    - cabot_someone_not_in_queue_bt_node
    - cabot_far_enough_bt_node
    - cabot_need_to_avoid_bt_node
    - cabot_need_to_replan_bt_node
    - cabot_check_path_bt_node
    - cabot_can_pass_person_bt_node
    - cabot_ignore_people_bt_node
    - cabot_person_stops_bt_node
    - cabot_change_param_bt_node
    - cabot_restore_config_bt_node
    - cabot_wait_float_bt_node
    - cabot_publish_topic_bt_node


bt_navigator_rclcpp_node:
  ros__parameters:
    use_sim_time: True


controller_server:
  ros__parameters:
    bond_timeout: 30.0
    bond_connect_timeout: 30.0
    bond_heartbeat_period: 1.0
    use_sim_time: True
    controller_frequency: 5.0
    min_x_velocity_threshold: 0.001
    min_y_velocity_threshold: 0.5
    min_theta_velocity_threshold: 0.001
    progress_checker_plugin: "progress_checker"
    goal_checker_plugins: ["cabot_goal_checker"]
    controller_plugins: ["FollowPath", "MPCFollowPath", "RLFollowPath"]

    # Progress checker parameters
    # It is not going to use progress_checker as of 2020/08/27.
    # It may need to identify this so that it can tell the user about the problem
    progress_checker:
      plugin: "nav2_controller::SimpleProgressChecker"
      required_movement_radius: 0.0 # disable checker (loalization's fractuation may enough
      movement_time_allowance: 10000000.0 # just for sure
    # Goal checker parameters
    cabot_goal_checker:
      plugin: "cabot_navigation2::CabotSimpleGoalChecker"
      xy_goal_tolerance: 0.50
      yaw_goal_tolerance: 6.28
      stateful: True
    # DWB parameters
    FollowPath:
      plugin: "dwb_core::DWBLocalPlanner"
      trajectory_generator_name: "dwb_plugins::StandardTrajectoryGenerator"
      debug_trajectory_details: True
    #  prune_plan: True # default: True
      prune_distance: 1.0 # default: 1.7 # should be greater than max_vel_x * sim_time + some margin
      min_vel_x: 0.0
      min_vel_y: 0.0
      max_vel_x: 0.5
      max_vel_y: 0.0
      max_vel_theta: 2.0
      min_speed_xy: 0.0
      max_speed_xy: 1.0
      min_speed_theta: 0.0
    # Add high threshold velocity for turtlebot 3 issue.
    # https://github.com/ROBOTIS-GIT/turtlebot3_simulations/issues/75
      acc_lim_x: 0.1
      acc_lim_y: 0.0
      acc_lim_theta: 1.0 # default 2.0
      decel_lim_x: -0.5
      decel_lim_y: 0.0
      decel_lim_theta: -1.0 # default -2.0
      vx_samples: 20
      vy_samples: 5
      vtheta_samples: 20
      sim_time: 1.7 #  default 1.7
    #  discretize_by_time: False # default = False
    #  time_granularity: 0.1 # default = 0.5
      linear_granularity: 0.05
      angular_granularity: 0.025
      transform_tolerance: 0.5
      #critics: ["Oscillation", "PathForwardPrefer", "BaseObstacle", "GoalAlign", "PathAlign", "PathDist", "GoalDist", "PoseAlign"]
      critics: ["Oscillation", "VelocityBaseObstacle", "GoalAlign", "PathAlign", "PathDist", "GoalDist"]
      PathForwardPrefer:
        class: dwb_critics::PathForwardPreferCritic
        scale: 24.0
        num_of_poses: 20
      VelocityBaseObstacle:
        class: dwb_critics::VelocityObstacleCritic
        #class: dwb_critics::BaseObstacleCritic
        scale: 1.0 # default: 1.0
        sum_scores: False # default: False
        low_speed_threshold: 0.9
        cost_threshold: 96.0
      PoseAlign:
        class: dwb_critics::PoseAlignCritic
        scale: 24.0
        minimum_angle: 135.0
      PathAlign:
        class: dwb_critics::PathAlignCritic
        scale: 24.0
      PathDist:
        class: dwb_critics::PathDistCritic
        scale: 24.0
      GoalAlign:
        class: dwb_critics::GoalAlignCritic
        forward_point_distance: 0.35
        scale: 16.0
      GoalDist:
        class: dwb_critics::GoalDistCritic
        scale: 16.0
      short_circuit_trajectory_evaluation: True
      lookahead_time: -1.0
      trans_stopped_velocity: 0.25
      slowing_factor: 5.0
      stateful: True
    MPCFollowPath:
      #plugin: "cabot_navigation2::CaBotRotationShimController"
      #primary_controller: "cabot_navigation2::CaBotSamplingMPCController"
      plugin: "cabot_navigation2::CaBotSamplingMPCController"
      group_topic: "/group_predictions"
      traj_vis_topic: "/local/control_output_vis"
      loc_goal_vis_topic: "/local/local_goal_vis"
      prediction_horizon: 4.8
      sampling_rate: 0.4
      max_linear_velocity: 0.5
      linear_sample_size: 4.0
      max_angular_velocity: 1.0
      angular_sample_size: 16.0
      obstacle_horizon: 2.0
      lookahead_distance: 2.5
      goal_distance: 1.0
      discount_factor: 0.99
      obstacle_costval: 250.0
      time_cost: 5.0
      focus_goal_dist: 2.0
      goal_cost_wt: 0.1
      group_cost_wt: 0.45
      obs_cost_wt: 1.0
      time_cost_wt: 1.0
      energy_cost_wt: 0.01
      smooth_wt: 0.0
    RLFollowPath:
      plugin: "cabot_navigation2::CaBotRLController"
      rl_cmd_topic: "/local/rl_robot_cmd"
      rl_info_topic: "/local/rl_robot_info"
      loc_goal_vis_topic: "/local/local_goal_vis"
      lookahead_distance: 2.5
      focus_goal_dist: 2.0

controller_server_rclcpp_node:
  ros__parameters:
    use_sim_time: True

local_costmap:
  local_costmap:
    ros__parameters:
      update_frequency: 5.0
      publish_frequency: 5.0
      global_frame: local/odom
      robot_base_frame: local/base_footprint
      use_sim_time: True
      rolling_window: True
      width: 10
      height: 10
      resolution: 0.05
      plugins: ["obstacle_layer", "low_obstacle_layer", "inflation_layer", "group_obstacle_layer"]
      robot_radius: 0.45
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        # robot_radius + safe_margin
        inflation_radius: 0.45
        cost_scaling_factor: 3.0
      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        enabled: True
        observation_sources: scan
        scan:
          topic: /local/scan
          min_obstacle_height: -100.0
          max_obstacle_height: 100.0
          obstacle_max_range: 4.0
          raytrace_max_range: 5.0
          clearing: True
          marking: True
          inf_is_valid: True # default: False
          data_type: "LaserScan"
      low_obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        enabled: True
        min_obstacle_height: -2.0
        max_obstacle_height: 2.0
        observation_sources: scan
        scan:
          topic: /local/livox_scan_expand
          min_obstacle_height: -2.0
          max_obstacle_height: 2.0
          obstacle_min_range: 0.05
          obstacle_max_range: 4.0
          raytrace_max_range: 5.0
          clearing: True
          marking: True
          inf_is_valid: True # default: False
          data_type: "LaserScan"
      group_obstacle_layer:
        plugin: "cabot_navigation2::GroupObstacleLayer"
        enabled: False
        group_topic: /group_predictions
        update_width: 10.0
        update_height: 10.0
        init_cost: 200.0
        discount_factor: 0.8
      always_send_full_costmap: True
  local_costmap_client:
    ros__parameters:
      use_sim_time: True
  local_costmap_rclcpp_node:
    ros__parameters:
      use_sim_time: True

global_costmap:
  global_costmap:
    ros__parameters:
      update_frequency: 1.0
      publish_frequency: 1.0
      robot_base_frame: local/base_footprint
      global_frame: local/odom
      rolling_window: True
      width: 20
      height: 20
      use_sim_time: True
      plugins: ["obstacle_layer", "low_obstacle_layer", "inflation_layer"]
      robot_radius: 0.45
      resolution: 0.05
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        # robot_radius + safe_margin
        inflation_radius: 0.35
        cost_scaling_factor: 5.0
      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        enabled: True
        observation_sources: scan
        scan:
          topic: /local/scan
          min_obstacle_height: -100.0
          max_obstacle_height: 100.0
          obstacle_max_range: 8.0
          raytrace_max_range: 8.0
          clearing: True
          marking: True
          inf_is_valid: True # default: False
          data_type: "LaserScan"
      low_obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        enabled: True
        min_obstacle_height: -2.0
        max_obstacle_height: 2.0
        observation_sources: scan
        scan:
          topic: /local/livox_scan_expand
          min_obstacle_height: -2.0
          max_obstacle_height: 2.0
          obstacle_min_range: 0.05
          obstacle_max_range: 4.5
          raytrace_max_range: 5.0
          clearing: True
          marking: True
          inf_is_valid: True # default: False
          data_type: "LaserScan"
      always_send_full_costmap: False
  global_costmap_client:
    ros__parameters:
      use_sim_time: True
  global_costmap_rclcpp_node:
    ros__parameters:
      use_sim_time: True

planner_server:
  ros__parameters:
    bond_timeout: 30.0
    bond_connect_timeout: 30.0
    bond_heartbeat_period: 1.0
    expected_planner_frequency: 2.0
    use_sim_time: True
    planner_plugins: ["GridBased", "PathForward"]
    GridBased:
      plugin: "nav2_navfn_planner/NavfnPlanner"
      tolerance: 0.5
      use_astar: false
      allow_unknown: false
    PathForward:
      plugin: "cabot_navigation2/CabotPathForwardPlanner"
      path_topic: "/path"

planner_server_rclcpp_node:
  ros__parameters:
    use_sim_time: True


behavior_server:
  ros__parameters:
    bond_timeout: 30.0
    bond_connect_timeout: 30.0
    bond_heartbeat_period: 1.0
    costmap_topic: /local/local_costmap/costmap_raw
    footprint_topic: /local/local_costmap/published_footprint
    cycle_frequency: 5.0
    behavior_plugins: ["spin", "backup", "wait"]
    spin:
      plugin: "nav2_behaviors/Spin"
    backup:
      plugin: "nav2_behaviors/BackUp"
    wait:
      plugin: "nav2_behaviors/Wait"
    global_frame: local/odom
    robot_base_frame: local/base_footprint
    transform_timeout: 0.1
    use_sim_time: true
    simulate_ahead_time: 2.0
    max_rotational_vel: 0.5
    min_rotational_vel: 0.01
    rotational_acc_lim: 0.2
